src/security/cli_security.c

```diff
@@ -161,3 +167,212 @@ get_cred_from_response(Drpc__Response *response, d_iov_t *cred)
+	if (rc != 0) {
+		D_ERROR("%s ACL is invalid\n", type_str);
+		return rc;
+	}
+
+	*acl_out = acl;
+	return 0;
+}
+
+static char *
+get_owner_str_from_prop(daos_prop_t *prop, uint32_t prop_type)
+{
+	struct daos_prop_entry	*entry;
+
+	D_ASSERT(prop_type == DAOS_PROP_PO_OWNER || prop_type == DAOS_PROP_PO_OWNER_GROUP ||
+		 prop_type == DAOS_PROP_CO_OWNER || prop_type == DAOS_PROP_CO_OWNER_GROUP);
+
+	entry = daos_prop_entry_get(prop, prop_type);
+	if (entry == NULL) {
+		D_ERROR("no entry for %d in property\n", prop_type);
+		return NULL;
+	}
+
+	return entry->dpe_str;
+}
+
+static int
+get_perms(daos_prop_t *prop, uint32_t acl_prop, uint32_t owner_prop, uint32_t group_prop,
+	  struct acl_user *user_info, uint64_t min_owner_perms, uint64_t *perms)
+{
+	struct daos_acl		*acl = NULL;
+	struct ownership	ownership = {0};
+	int			rc;
+
+	/* These ACL and ownership variables point to the data in-place in the prop, and thus don't
+	 * need to be freed here.
+	 */
+	rc = acl_from_prop(prop, acl_prop, &acl);
+	if (rc != 0)
+		return rc;
+
+	ownership.user = get_owner_str_from_prop(prop, owner_prop);
+	if (ownership.user == NULL) {
+		D_ERROR("couldn't get owner user (%d) from prop\n", owner_prop);
+		return -DER_INVAL;
+	}
+
+	ownership.group = get_owner_str_from_prop(prop, group_prop);
+	if (ownership.group == NULL) {
+		D_ERROR("couldn't get owner group (%d) from prop\n", group_prop);
+		return -DER_INVAL;
+	}
+
+	return get_acl_permissions(acl, &ownership, user_info, min_owner_perms, perms);
+}
+
+static void
+free_user_info_strings(struct acl_user *user_info)
+{
+	size_t i;
+
+	for (i = 0; i < user_info->nr_groups; i++)
+		D_FREE(user_info->groups[i]);
+	D_FREE(user_info->groups);
+	D_FREE(user_info->group);
+	D_FREE(user_info->user);
+}
+
+static int
+fill_user_info(uid_t uid, gid_t gid, gid_t *gids, size_t nr_gids, struct acl_user *user_info)
+{
+	int	rc;
+	size_t	i;
+
+	rc = daos_acl_uid_to_principal(uid, &user_info->user);
+	if (rc != 0) {
+		D_ERROR("failed to convert uid %d to an ACL principal: "DF_RC"\n", uid, DP_RC(rc));
+		D_GOTO(err_out, rc);
+	}
+
+	rc = daos_acl_gid_to_principal(gid, &user_info->group);
+	if (rc != 0) {
+		D_ERROR("failed to convert gid %d to an ACL principal: "DF_RC"\n", gid, DP_RC(rc));
+		D_GOTO(err_user, rc);
+	}
+
+	D_ALLOC_ARRAY(user_info->groups, nr_gids);
+	if (user_info->groups == NULL)
+		D_GOTO(err_group, rc = -DER_NOMEM);
+
+	user_info->nr_groups = 0;
+	for (i = 0; i < nr_gids; i++) {
+		rc = daos_acl_gid_to_principal(gids[i], &user_info->groups[i]);
+		if (rc != 0) {
+			D_ERROR("failed to convert gid %d to an ACL principal: "DF_RC"\n", gids[i],
+				DP_RC(rc));
+			D_GOTO(err_grps, rc);
+		}
+		user_info->nr_groups++;
+	}
+	D_ASSERT(nr_gids == user_info->nr_groups);
+
+	return rc;
+
+err_grps:
+	for (i = 0; i < user_info->nr_groups; i++)
+		D_FREE(user_info->groups[i]);
+	D_FREE(user_info->groups);
+err_group:
+	D_FREE(user_info->group);
+err_user:
+	D_FREE(user_info->user);
+err_out:
```

Comment:
You could probably just simplify all this by just calling free_user_info_strings().

Reply:
Fixed